File: src/eko/tools/toolList/write-to-file.ts
/**
 * write_to_file 工具
 * 支持：
 * - 覆盖写入或新建文件
 */

import { ModifiableTool, type ToolExecutionContext } from '../modifiable-tool'
import type { ToolResult } from '@/eko-core/types'
import { filesystemApi } from '@/api'
import { ValidationError, ToolError } from '../tool-error'
import { writeTextFile } from '@tauri-apps/plugin-fs'

export interface WriteToFileParams {
  path: string
  content: string
}

export class WriteToFileTool extends ModifiableTool {
  constructor() {
    super(
      'write_to_file',
      `Create or overwrite a file with provided content.
Notes:
- Absolute paths are required.
- If the file exists, it will be overwritten entirely.`,
      {
        type: 'object',
        properties: {
          path: { type: 'string', description: 'Absolute file path' },
          content: { type: 'string', description: 'New file content' },
        },
        required: ['path', 'content'],
      }
    )
  }

  protected async executeImpl(context: ToolExecutionContext): Promise<ToolResult> {
    const params = context.parameters as unknown as WriteToFileParams

    const path = (params.path || '').toString().trim()
    if (!path) throw new ValidationError('File path cannot be empty')
    if (!isAbsolutePath(path)) throw new ValidationError('Path must be an absolute path')

    const content = params.content ?? ''

    const exists = await filesystemApi.exists(path)
    const isDir = exists ? await filesystemApi.isDirectory(path) : false
    if (isDir) throw new ValidationError(`Path ${path} is a directory, cannot write file`)
    if (isBinaryFile(path)) throw new ValidationError(`File ${path} appears to be binary, text writing not supported`)

    const created = !exists

    try {
      await writeTextFile(path, content)
    } catch (e) {
      throw new ToolError(`Failed to write file: ${e instanceof Error ? e.message : String(e)}`)
    }

    return {
      content: [
        {
          type: 'text',
          text: `write_to_file applied\nFile: ${path}\nCreated: ${created}`,
        },
      ],
      extInfo: {
        file: path,
        created,
      },
    }
  }
}

export const writeToFileTool = new WriteToFileTool()

// ===== 辅助函数 =====
function isAbsolutePath(p: string): boolean {
  return p.startsWith('/')
}

function isBinaryFile(path: string): boolean {
  const ext = path.toLowerCase().split('.').pop() || ''
  const binaryExtensions = new Set([
    'jpg',
    'jpeg',
    'png',
    'gif',
    'bmp',
    'tiff',
    'tif',
    'webp',
    'ico',
    'svg',
    'mp3',
    'wav',
    'flac',
    'aac',
    'ogg',
    'm4a',
    'wma',
    'mp4',
    'avi',
    'mkv',
    'mov',
    'wmv',
    'flv',
    'webm',
    '3gp',
    'zip',
    'rar',
    '7z',
    'tar',
    'gz',
    'bz2',
    'xz',
    'exe',
    'dll',
    'so',
    'dylib',
    'app',
    'deb',
    'rpm',
    'dmg',
    'doc',
    'docx',
    'xls',
    'xlsx',
    'ppt',
    'pptx',
    'pdf',
    'ttf',
    'otf',
    'woff',
    'woff2',
    'eot',
    'db',
    'sqlite',
    'sqlite3',
    'class',
    'jar',
    'war',
    'ear',
    'pyc',
    'pyo',
    'o',
    'obj',
    'bin',
    'dat',
    'iso',
    'img',
  ])
  return binaryExtensions.has(ext)
}

File: src-tauri/src/agent/tools/builtin/write_to_file.rs
use std::path::PathBuf;

use async_trait::async_trait;
use serde::Deserialize;
use serde_json::json;
use tokio::fs;

use crate::agent::state::context::TaskContext;
use crate::agent::tools::{
    RunnableTool, ToolExecutorResult, ToolPermission, ToolResult, ToolResultContent,
};

use super::file_utils::{ensure_absolute, is_probably_binary};

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct WriteToFileArgs {
    path: String,
    content: String,
}

pub struct WriteToFileTool;

impl WriteToFileTool {
    pub fn new() -> Self {
        Self
    }
}

#[async_trait]
impl RunnableTool for WriteToFileTool {
    fn name(&self) -> &str {
        "write_to_file"
    }

    fn description(&self) -> &str {
        "Create or overwrite a file with provided text content."
    }

    fn parameters_schema(&self) -> serde_json::Value {
        json!({
            "type": "object",
            "properties": {
                "path": {
                    "type": "string",
                    "description": "Absolute file path"
                },
                "content": {
                    "type": "string",
                    "description": "New file content"
                }
            },
            "required": ["path", "content"]
        })
    }

    fn required_permissions(&self) -> Vec<ToolPermission> {
        vec![ToolPermission::FileSystem]
    }

    fn tags(&self) -> Vec<String> {
        vec!["file".into(), "write".into(), "overwrite".into()]
    }

    async fn run(
        &self,
        _context: &TaskContext,
        args: serde_json::Value,
    ) -> ToolExecutorResult<ToolResult> {
        let args: WriteToFileArgs = serde_json::from_value(args)?;
        let path = PathBuf::from(&args.path);

        if let Err(msg) = ensure_absolute(&path) {
            return Ok(validation_error(msg));
        }

        if is_probably_binary(&path) {
            return Ok(validation_error(format!(
                "File {} appears to be binary, text writing not supported",
                path.display()
            )));
        }

        if let Some(parent) = path.parent() {
            if !parent.exists() {
                return Ok(tool_error(format!(
                    "Parent directory does not exist: {}",
                    parent.display()
                )));
            }
        }

        let (created, is_directory) = match fs::metadata(&path).await {
            Ok(meta) => (false, meta.is_dir()),
            Err(_) => (true, false),
        };

        if is_directory {
            return Ok(validation_error(format!(
                "Path {} is a directory, cannot write file",
                path.display()
            )));
        }

        if let Err(err) = fs::write(&path, args.content).await {
            return Ok(tool_error(format!(
                "Failed to write file {}: {}",
                path.display(),
                err
            )));
        }

        let message = format!(
            "write_to_file applied\nFile: {}\nCreated: {}",
            path.display(),
            created
        );

        Ok(ToolResult {
            content: vec![ToolResultContent::Text { text: message }],
            is_error: false,
            execution_time_ms: None,
            metadata: Some(json!({
                "file": path.display().to_string(),
                "created": created,
            })),
        })
    }
}

fn validation_error(message: impl Into<String>) -> ToolResult {
    ToolResult {
        content: vec![ToolResultContent::Error {
            message: message.into(),
            details: None,
        }],
        is_error: true,
        execution_time_ms: None,
        metadata: None,
    }
}

fn tool_error(message: impl Into<String>) -> ToolResult {
    ToolResult {
        content: vec![ToolResultContent::Error {
            message: message.into(),
            details: None,
        }],
        is_error: true,
        execution_time_ms: None,
        metadata: None,
    }
}
